pkg/kubelet/kubelet.go

```diff
@@ -2313,6 +2376,99 @@ func (kl *Kubelet) HandlePodSyncs(pods []*v1.Pod) {
 	}
 }
 
+func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, *v1.Pod, v1.PodResizeStatus) {
+	var otherActivePods []*v1.Pod
+
+	node, err := kl.getNodeAnyWay()
+	if err != nil {
+		klog.ErrorS(err, "getNodeAnyway function failed")
+		return false, nil, ""
+	}
+	cpuAvailable := node.Status.Allocatable.Cpu().MilliValue()
+	memAvailable := node.Status.Allocatable.Memory().Value()
+	cpuRequests := resource.GetResourceRequest(pod, v1.ResourceCPU)
+	memRequests := resource.GetResourceRequest(pod, v1.ResourceMemory)
+	if cpuRequests > cpuAvailable || memRequests > memAvailable {
+		klog.V(2).InfoS("Resize is not feasible as request exceeds available node resources", "Pod", pod.Name)
+		return false, nil, v1.PodResizeStatusInfeasible
+	}
+
+	// Treat the existing pod needing resize as a new pod with desired resources seeking admit.
+	// If desired resources don't fit, pod continues to run with currently allocated resources.
+	activePods := kl.GetActivePods()
+	for _, p := range activePods {
+		if p.UID != pod.UID {
+			otherActivePods = append(otherActivePods, p)
+		}
+	}
+
+	if ok, failReason, failMessage := kl.canAdmitPod(otherActivePods, pod); !ok {
+		// Log reason and return. Let the next sync iteration retry the resize
+		klog.V(2).InfoS("Resize cannot be accommodated", "Pod", pod.Name, "Reason", failReason, "Message", failMessage)
+		return false, nil, v1.PodResizeStatusDeferred
+	}
+
+	podCopy := pod.DeepCopy()
+	for _, container := range podCopy.Spec.Containers {
+		idx, _, _ := podutil.GetContainerStatus(podCopy.Status.ContainerStatuses, container.Name)
+		for rName, rQuantity := range container.Resources.Requests {
+			podCopy.Status.ContainerStatuses[idx].ResourcesAllocated[rName] = rQuantity
+		}
+	}
+	return true, podCopy, v1.PodResizeStatusInProgress
+}
+
+func (kl *Kubelet) handlePodResourcesResize(pod *v1.Pod) {
+	if pod.Status.Phase != v1.PodRunning {
+		return
+	}
+	podResized := false
+	for _, container := range pod.Spec.Containers {
+		if len(container.Resources.Requests) == 0 {
+			continue
+		}
+		_, containerStatus, found := podutil.GetContainerStatus(pod.Status.ContainerStatuses, container.Name)
+		if !found {
+			klog.V(3).InfoS("ContainerStatus not found", "pod", pod.Name, "container", container.Name)
```

Comment:
This should be V=5, this will cause a ton of logspam as written.

Reply:
Fixed in latest commit, please review.